File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/PropostaService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.model.CampoTipoProposta;
import br.com.portalproposta.backend.model.Proposta;
import br.com.portalproposta.backend.model.TipoProposta;
import br.com.portalproposta.backend.repository.ClienteRepository;
import br.com.portalproposta.backend.repository.EmpresaRepository;
import br.com.portalproposta.backend.repository.PropostaRepository;
import br.com.portalproposta.backend.repository.TipoPropostaRepository;
import tools.jackson.databind.JsonNode;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class PropostaService {

    private final PropostaRepository propostaRepository;
    private final ClienteRepository clienteRepository;
    private final TipoPropostaRepository tipoPropostaRepository;
    private final EmpresaRepository empresaRepository;
    private final ConfiguracaoService configuracaoService;

    @Autowired
    public PropostaService(PropostaRepository propostaRepository,
            ClienteRepository clienteRepository,
            TipoPropostaRepository tipoPropostaRepository,
            EmpresaRepository empresaRepository,
            ConfiguracaoService configuracaoService) {
        this.propostaRepository = propostaRepository;
        this.clienteRepository = clienteRepository;
        this.tipoPropostaRepository = tipoPropostaRepository;
        this.empresaRepository = empresaRepository;
        this.configuracaoService = configuracaoService;
    }

    public List<Proposta> list() {
        return propostaRepository.findAll();
    }

    public List<Proposta> listByClienteId(String clienteId) {
        if (!clienteRepository.existsById(clienteId)) {
            throw new AppNotFoundException("cliente não encontrado");
        }
        return propostaRepository.findByClienteId(clienteId);
    }

    public Proposta getById(String id) {
        return propostaRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("proposta não encontrada"));
    }

    @Transactional
    public Proposta create(Proposta proposal) {
        if (!clienteRepository.existsById(proposal.getClienteId())) {
            throw new AppNotFoundException("cliente não encontrado");
        }

        TipoProposta tipoProposta = tipoPropostaRepository.findByChave(proposal.getTipo())
                .orElseThrow(() -> new AppBadRequestException(
                        "tipo de proposta inválido (use Imobiliaria, Auto ou Comissionados)"));

        if (proposal.getValor() <= 0) {
            throw new AppBadRequestException("valor da proposta deve ser maior que zero");
        }

        validateDynamicFields(tipoProposta, proposal.getDadosEspecificos());

        proposal.setId(UUID.randomUUID().toString());
        if (proposal.getStatus() == null || proposal.getStatus().trim().isEmpty()) {
            proposal.setStatus("Pendente");
        }

        handleBrokerageCommission(proposal);

        proposal.setCreatedAt(LocalDateTime.now());
        proposal.setUpdatedAt(LocalDateTime.now());

        Proposta saved = propostaRepository.save(proposal);
        loadRelations(saved);
        return saved;
    }

    @Transactional
    public Proposta update(String id, Proposta input) {
        Proposta proposal = getById(id);

        if (input.getTipo() != null && !input.getTipo().trim().isEmpty()) {
            TipoProposta tipoProposta = tipoPropostaRepository.findByChave(input.getTipo())
                    .orElseThrow(() -> new AppBadRequestException(
                            "tipo de proposta inválido (use Imobiliaria, Auto ou Comissionados)"));

            validateDynamicFields(tipoProposta, input.getDadosEspecificos());
            proposal.setTipo(input.getTipo());
        }

        if (input.getValor() <= 0) {
            throw new AppBadRequestException("valor da proposta deve ser maior que zero");
        }

        proposal.setValor(input.getValor());
        proposal.setStatus(input.getStatus());
        proposal.setDescricao(input.getDescricao());
        proposal.setDadosEspecificos(input.getDadosEspecificos());
        proposal.setEmpresaId(input.getEmpresaId());
        proposal.setStatusCorretagem(input.getStatusCorretagem());

        handleBrokerageCommission(proposal);

        proposal.setUpdatedAt(LocalDateTime.now());

        Proposta saved = propostaRepository.save(proposal);
        loadRelations(saved);
        return saved;
    }

    @Transactional
    public void delete(String id) {
        Proposta proposal = getById(id);
        propostaRepository.delete(proposal);
    }

    public void handleBrokerageCommission(Proposta p) {
        if (p.getEmpresaId() != null
                && ("FechadaComSucesso".equals(p.getStatusCorretagem()) || "Aprovada".equals(p.getStatus()))) {
            p.setStatusCorretagem("FechadaComSucesso");
            double taxa = configuracaoService.getCommissionRate();
            double value = p.getValor() * (taxa / 100.0);
            BigDecimal decimalValue = new BigDecimal(value).setScale(2, RoundingMode.HALF_UP);
            p.setValorComissao(decimalValue.doubleValue());
        } else if (p.getEmpresaId() != null) {
            p.setStatusCorretagem("Encaminhada");
            p.setValorComissao(0);
        } else {
            p.setStatusCorretagem("Pendente");
            p.setValorComissao(0);
        }
    }

    private void loadRelations(Proposta p) {
        if (p.getEmpresaId() != null) {
            p.setEmpresa(empresaRepository.findById(p.getEmpresaId()).orElse(null));
        } else {
            p.setEmpresa(null);
        }
    }

    private void validateDynamicFields(TipoProposta tipo, JsonNode dadosEspecificos) {
        if ("Imobiliaria".equals(tipo.getChave()) || "Auto".equals(tipo.getChave())
                || "Comissionados".equals(tipo.getChave())) {
            return;
        }

        boolean hasData = dadosEspecificos != null && dadosEspecificos.isObject();

        for (CampoTipoProposta campo : tipo.getCampos()) {
            boolean exists = hasData && dadosEspecificos.has(campo.getChave());
            JsonNode propValue = exists ? dadosEspecificos.get(campo.getChave()) : null;

            if (campo.isObrigatorio()) {
                if (propValue == null || propValue.isNull() || propValue.isMissingNode()) {
                    throw new AppBadRequestException(String.format("o campo '%s' é obrigatório", campo.getNome()));
                }
                if (propValue.isTextual() && propValue.asText().trim().isEmpty()) {
                    throw new AppBadRequestException(String.format("o campo '%s' é obrigatório", campo.getNome()));
                }
            }

            if (exists && propValue != null && !propValue.isNull() && !propValue.isMissingNode()) {
                if ("number".equals(campo.getTipo())) {
                    if (!propValue.isNumber()) {
                        if (propValue.isTextual()) {
                            try {
                                Double.parseDouble(propValue.asText());
                            } catch (NumberFormatException e) {
                                throw new AppBadRequestException(
                                        String.format("o campo '%s' deve ser um número", campo.getNome()));
                            }
                        } else {
                            throw new AppBadRequestException(
                                    String.format("o campo '%s' deve ser um número", campo.getNome()));
                        }
                    }
                } else if ("boolean".equals(campo.getTipo())) {
                    if (!propValue.isBoolean()) {
                        if (propValue.isTextual()) {
                            String text = propValue.asText();
                            if (!"true".equals(text) && !"false".equals(text)) {
                                throw new AppBadRequestException(
                                        String.format("o campo '%s' deve ser verdadeiro ou falso", campo.getNome()));
                            }
                        } else {
                            throw new AppBadRequestException(
                                    String.format("o campo '%s' deve ser verdadeiro ou falso", campo.getNome()));
                        }
                    }
                }
            }
        }
    }
}

File: backend/Services/PropostaService.cs
using System.Globalization;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using PortalProposta.Data;
using PortalProposta.Models;

namespace PortalProposta.Services;

public class PropostaService(AppDbContext context)
{
    public async Task<List<Proposta>> ListAsync() =>
        await context.Propostas.Include(p => p.Empresa).ToListAsync();

    public async Task<List<Proposta>> ListByClienteIdAsync(string clienteId)
    {
        if (!await context.Clientes.AnyAsync(c => c.Id == clienteId))
            throw new AppNotFoundException("cliente não encontrado");

        return await context.Propostas
            .Include(p => p.Empresa)
            .Where(p => p.ClienteId == clienteId)
            .ToListAsync();
    }

    public async Task<Proposta> GetByIdAsync(string id) =>
        await context.Propostas.Include(p => p.Empresa).FirstOrDefaultAsync(p => p.Id == id)
            ?? throw new AppNotFoundException("proposta não encontrada");

    public async Task<Proposta> CreateAsync(Proposta proposal)
    {
        if (!await context.Clientes.AnyAsync(c => c.Id == proposal.ClienteId))
            throw new AppNotFoundException("cliente não encontrado");

        var tipoProposta = await context.TiposProposta.FirstOrDefaultAsync(t => t.Chave == proposal.Tipo)
            ?? throw new AppBadRequestException("tipo de proposta inválido (use Imobiliaria, Auto ou Comissionados)");

        if (proposal.Valor <= 0)
            throw new AppBadRequestException("valor da proposta deve ser maior que zero");

        ValidateDynamicFields(tipoProposta, proposal.DadosEspecificos);

        proposal.Id = Guid.NewGuid().ToString();
        if (string.IsNullOrWhiteSpace(proposal.Status))
            proposal.Status = "Pendente";

        await HandleBrokerageCommissionAsync(proposal);
        proposal.CreatedAt = proposal.UpdatedAt = DateTime.Now;

        context.Propostas.Add(proposal);
        await context.SaveChangesAsync();
        return proposal;
    }

    public async Task<Proposta> UpdateAsync(string id, Proposta input)
    {
        var proposal = await context.Propostas.FirstOrDefaultAsync(p => p.Id == id)
            ?? throw new AppNotFoundException("proposta não encontrada");

        if (!string.IsNullOrWhiteSpace(input.Tipo))
        {
            var tipoProposta = await context.TiposProposta.FirstOrDefaultAsync(t => t.Chave == input.Tipo)
                ?? throw new AppBadRequestException("tipo de proposta inválido");

            ValidateDynamicFields(tipoProposta, input.DadosEspecificos);
            proposal.Tipo = input.Tipo;
        }

        if (input.Valor <= 0)
            throw new AppBadRequestException("valor da proposta deve ser maior que zero");

        proposal.Valor = input.Valor;
        proposal.Status = input.Status;
        proposal.Descricao = input.Descricao;
        proposal.DadosEspecificos = input.DadosEspecificos;
        proposal.EmpresaId = input.EmpresaId;
        proposal.StatusCorretagem = input.StatusCorretagem;

        await HandleBrokerageCommissionAsync(proposal);
        proposal.UpdatedAt = DateTime.Now;

        await context.SaveChangesAsync();

        // Reload empresa association
        proposal.Empresa = proposal.EmpresaId is not null
            ? await context.Empresas.FirstOrDefaultAsync(e => e.Id == proposal.EmpresaId)
            : null;

        return proposal;
    }

    public async Task DeleteAsync(string id)
    {
        var proposal = await context.Propostas.FirstOrDefaultAsync(p => p.Id == id)
            ?? throw new AppNotFoundException("proposta não encontrada");

        context.Propostas.Remove(proposal);
        await context.SaveChangesAsync();
    }

    public async Task HandleBrokerageCommissionAsync(Proposta p)
    {
        if (p.EmpresaId is not null && (p.StatusCorretagem == "FechadaComSucesso" || p.Status == "Aprovada"))
        {
            p.StatusCorretagem = "FechadaComSucesso";
            var taxa = await GetCommissionRateAsync();
            p.ValorComissao = Math.Round(p.Valor * (taxa / 100.0), 2);
        }
        else if (p.EmpresaId is not null)
        {
            p.StatusCorretagem = "Encaminhada";
            p.ValorComissao = 0;
        }
        else
        {
            p.StatusCorretagem = "Pendente";
            p.ValorComissao = 0;
        }
    }

    private async Task<double> GetCommissionRateAsync()
    {
        var conf = await context.Configuracoes.FirstOrDefaultAsync(c => c.Chave == "taxa_corretagem");
        return conf is not null && double.TryParse(conf.Valor, NumberStyles.Any, CultureInfo.InvariantCulture, out var rate)
            ? rate
            : 5.0;
    }

    private static void ValidateDynamicFields(TipoProposta tipo, JsonElement dadosEspecificos)
    {
        if (tipo.Chave is "Imobiliaria" or "Auto" or "Comissionados")
            return;

        var hasData = dadosEspecificos.ValueKind == JsonValueKind.Object;

        foreach (var campo in tipo.Campos)
        {
            if (!hasData)
            {
                if (campo.Obrigatorio)
                    throw new AppBadRequestException($"o campo '{campo.Nome}' é obrigatório");
                continue;
            }

            var exists = dadosEspecificos.TryGetProperty(campo.Chave, out var propValue);

            if (campo.Obrigatorio)
            {
                if (!exists || propValue.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
                    throw new AppBadRequestException($"o campo '{campo.Nome}' é obrigatório");

                if (propValue.ValueKind == JsonValueKind.String && string.IsNullOrWhiteSpace(propValue.GetString()))
                    throw new AppBadRequestException($"o campo '{campo.Nome}' é obrigatório");
            }

            if (!exists || propValue.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
                continue;

            switch (campo.Tipo)
            {
                case "number" when propValue.ValueKind != JsonValueKind.Number:
                {
                    if (propValue.ValueKind == JsonValueKind.String &&
                        double.TryParse(propValue.GetString(), NumberStyles.Any, CultureInfo.InvariantCulture, out _))
                        break;
                    throw new AppBadRequestException($"o campo '{campo.Nome}' deve ser um número");
                }
                case "boolean" when propValue.ValueKind is not JsonValueKind.True and not JsonValueKind.False:
                {
                    if (propValue.ValueKind == JsonValueKind.String)
                    {
                        var strVal = propValue.GetString();
                        if (strVal is "true" or "false") break;
                    }
                    throw new AppBadRequestException($"o campo '{campo.Nome}' deve ser verdadeiro ou falso");
                }
            }
        }
    }
}
